openage/convert/opus/opusenc.pyx

```diff
@@ -0,0 +1,428 @@
+        PyMem_Free(pymem2)
+    return
+
+
+cdef upsample(indata, dict opts, int target_rate):
+    '''
+    Upsamples the PCM-data in 'indata' to 'target_rate' using linear
+    interpolation. 'opts' is a dict containing info on 'indata'.
+    If successful, return the upsampled data, otherwise an error message.
+    Assumptions: * 1 ≤ opts['bit_depth'] ≤ 16
+                 * opts['input_rate'] < target_rate
+                 * opts['channels'] ∈ {1, 2}
+    TODO: * a more eloquent version?
+          * a faster version?
+          * a better interpolation?
+    '''
+    cdef:
+        int wframe_len = opts['wframe_len']
+        int alpha, ch, channel = opts['channels']
+        int bit_depth = opts['bits_per_sample']
+        unsigned short mask = ~((1 << (16 - bit_depth)) - 1)
+
+        size_t iin = 0, iout = 0
+        size_t ismpls, osmpls  # number of input and output samples
+        short a,b
+        int num, den, divisor
+
+        size_t outsize
+        const char *inp = indata
+        char *out
+
+    ismpls = len(indata) // wframe_len
+    if ismpls < 2:
+        # Nothing to do.
+        return indata
+
+    divisor = gcd(target_rate, opts['input_rate'])
+    num = <int> (target_rate / divisor)
+    den = <int> (opts['input_rate'] / divisor)
+
+    # We're using linear interpolation. That means, we need for every output
+    # sample one input samples that comes before and one after the output
+    # sample. ismpls - 1 is the index of the last input sample. By multiplying
+    # with num / den we get the index of the last output sample. In general,
+    # this is not a whole number. Therefore we take only the integer part.
+    # Because indices are zero-based, add one for the number of output samples.
+    osmpls = ((ismpls - 1) * num / den).__trunc__() + 1
+    outsize = osmpls * channel * 2  # we convert to int16
+
+    try:
+        ret = bytearray(outsize)
+    except MemoryError:
+        return "Could not allocate buffer."
+
+    out = PyByteArray_AS_STRING(ret)
+
+    # iout is the next sample we want to calculate. We need one sample earlier
+    # and one later in time. Those are at indices iin and iin + 1. We need to
+    # check in every iteration, if iout is now after iin + 1 and in that case
+    # increase iin by one. Because we are _up_-sampling, the time between two
+    # output samples is shorter than between two input samples, and therefore
+    # adding one to iin suffices to keep the invariant
+    #           iin * num/den ≤ iout ≤ (iin + 1) * num/den .
+    if bit_depth > 8:
+        for iout in range(0, osmpls):
+            if (iin + 1) * num < iout * den:
+                iin += 1
+
+            alpha = iout * den / num - iin
+            for ch in range(0, channel):
+                a = (<short *> inp)[iin * channel + ch]
+                b = (<short *> inp)[(iin + 1) * channel + ch]
+                (<short *> out)[(iout + ch) * channel] = (<short> (a + alpha * (b - a))) & mask
```

Comment:
those casts are always 16 bit, right? a `uint16_t` cast might be more platform compatible.